### ISSUE TITLE
[Bug]: page.waitForNetworkIdle timeout inconsistently 

### ISSUE THREAD

**Issue by Josehower** [puppeteer/puppeteer#8383]
### Bug description
For some reason `page.waitForNetworkIdle` is timing out for no reason. the network is empty and even so it timeout after 30 seconds.

This error happens only sometimes and is very weird. Despite of that, I find a way to make it fail consistently by repeating some steps multiple times

### Steps to reproduce the problem:

I created a repo with some simple [puppeteer code ](https://github.com/Josehower/puppetter-timeout-reproduction/blob/main/index.ts)

### Locally 
you can reproduce it locally by (my current local environment is WSL2 ubuntu)
1. clone the repo
2. install dependencies `yarn`
3. `yarn start`

### In GitHub Actions
 1. Create a new issue in the [reproduction repo ](https://github.com/Josehower/puppetter-timeout-reproduction/blob/main/index.ts) or use the issue `test`
 2. Create any comment
 3. Check the actions tab

### Puppeteer version

  "puppeteer": "14.1.1",

### Node.js version

Local v16.15.0  / [node 16](https://github.com/Josehower/puppetter-timeout-reproduction/blob/8be56524cbfbfb1d71561b25f4b6c0c283a41b94/.github/workflows/upleveled-drone.yml#L18)

### npm version

local 8.9.0 / actions https://github.com/actions/checkout

### What operating system are you seeing the problem on?

Linux

### Relevant log output

https://github.com/Josehower/puppetter-timeout-reproduction/runs/6524437119?check_suite_focus=true#step:4:20

```shell
file:///home/runner/work/puppetter-timeout-reproduction/puppetter-timeout-reproduction/upleveled-drone.mjs:18421
      rejectCallback(new TimeoutError("Timeout exceeded while waiting for event"));
                     ^

TimeoutError: Timeout exceeded while waiting for event
    at Timeout.<anonymous> (file:///home/runner/work/puppetter-timeout-reproduction/puppetter-timeout-reproduction/upleveled-drone.mjs:18421:22)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
```

Thanks for any attention that you gave to this issue, puppeteer is an amazing tool that we use every day in my company. I am also open to creating a PR if we are able to identify what is the problem.

Good energy 👍👍👍👍

**Comment by OrKoN** [comment]
Yeah, looks like some sort of bug! Thanks for reporting.

**Comment by Josehower** [comment]
Adding some context to where the app times out.

The app is tested by `puppeteer` updating an image source with a nonvalid URL, which results in a 404 response from the server, the image update is happening on each letter type which causes a bunch of 404 happening quick.

![image](https://user-images.githubusercontent.com/61600906/169981674-dad097d3-d6f3-4c47-aac9-db6c65375dec.png)

The code that fails is 
```js
  await element.type('lol');

  console.log('network test start...');
  await page.waitForNetworkIdle();
  await console.log('network check 1 pass');

  await element.type('lol');
  await page.waitForNetworkIdle();
  console.log('network check 2 pass');

  await element.type('lol');
  await page.waitForNetworkIdle();
  console.log('network check 3 pass');
  
  .... more repetition of type + waitForNetworkIdle
```

The unexpected behavior is sometimes this `page.waitForNetworkIdle()` is timing out after 30 secs which is very weird. also, the block that fails is always different. (sometimes fails in check No1 and sometimes breaks on check 6 etc)

The problem is this error is not always happening, actually, most of the `waitForNetworkIdle`'s do not timeout what would be the expected behavior, since a broken link should respond with a 404 very quick.

Add again here a link to the[ full GitHub actions workflow with the fail example](https://github.com/Josehower/puppetter-timeout-reproduction/runs/6524437119?check_suite_focus=true#step:4:20)

**Comment by karlhorky** [comment]
@OrKoN did you have a chance to take a look at this?

Or do you have any tips as to how @Josehower can look through the Puppeteer codebase to find the bug?

**Comment by OrKoN** [comment]
Hey, didn't have time to look into that yet. The code for the function is here https://github.com/puppeteer/puppeteer/blob/main/src/common/Page.ts#L1944 It looks like sometimes the request count does not update as expected.

**Comment by chaoticsparks** [comment]
Is there any update on that?

**Comment by karlhorky** [comment]
I'm guessing this is not stale - but maybe the `confirmed` label already makes the stale bot go away...

**Comment by sagrawal31** [comment]
For me, this was happening very randomly when I was trying to take 25 screenshots in parallel. To investigate this, I made the screenshot generation sequential and then I figured out that because of a CORS issue (in loading a font), the `waitForNetworkIdle` never finishes and always gets timed out (irrespective of whether it was set for a minute).

Hope this helps in reproducing the issue.

**Comment by gomain** [comment]
Having similar issue here. I notice that when an ajax request completes very quickly I get this timeout. As if the monitoring of network being idle begins after the network is already idle.

**Comment by roths** [comment]
> We're marking this issue as unconfirmed because it has not had recent activity and we weren't able to confirm it yet. It will be closed if no further activity occurs within the next 30 days.

exactly，i have the same issue

**Comment by kiber-io** [comment]
I ran into the same problem and I think I found out why this is happening.

In my case (you may have something else), the site sends [**beacon** requests](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/ResourceType#beacon), which chrome marks as **ping** (judging by the [mdn documentation](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/ResourceType#ping)):
> Browsers that don't have a dedicated beacon type (see browser compatibility below), also label requests sent through the Beacon API as ping.

_(in the compatibility table, chrome does not support beacon)_

Beacon requests, in turn, do not require responses, which is why when checking in the [NetworkEventManager.numRequestsInProgress](https://github.com/puppeteer/puppeteer/blob/f42336cf83982332829ca7e14ee48d8676e11545/packages/puppeteer-core/src/common/NetworkEventManager.ts#L153) **request.response() is always null** and puppeteer believes that the request has not yet been completed.

The solution is either to add a check to NetworkEventManager.numRequestsInProgress on request.resourceType() === 'ping' or not to add this type to httpRequestsMap at all.

Something like this
```js
storeRequest(networkRequestId: NetworkRequestId, request: HTTPRequest): void {
    // Add this line for check ping/beacon type
    if (request.resourceType() === 'ping') return;

    this.#httpRequestsMap.set(networkRequestId, request);
}
```

**Comment by ProchaLu** [comment]
Maybe also a weird behavior. The test fails less frequently when running the code inside a loop and executing multiple times in a row. Maybe there is less idle time between each execution, which could cause the network to become idle faster. This could explain why the code inside the loop fails less frequently than the code without the loop.

Issue:

- https://github.com/puppeteer/puppeteer/issues/9881

**Comment by jpgklassen** [comment]
I am having the same issue!

**Comment by janwirth** [comment]
`idleTime: 100` seems to improve this for me, I was running page.goTo and waitForNetworkIdle against a local HTML server.

Edit: false alarm. It doesn't work. Also adding server timeout doesn't work.

I'm using bun.serve
